### PR TITLE
[TRTLLM-1302][feat] Topk logprobs for TRT backend and top1 logprob for PyT backend

### PR DESCRIPTION
**Pull request by LinPoly**
## Description

1. TopK logprobs for TRT backend with gathering context & generation logits
2. Top1 logprobs for PyT backend
For now some length mismatch happens randomly, still need to investigate.
## Test Coverage

Will be added.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--disable-fail-fast --skip-test --stage-list "A10-1, xxx" --gpu-type "A30, H100_PCIe" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-[Post-Merge]-1, xxx"]`

Launch build/test pipelines. All previously running jobs will be killed.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests. Will also run L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-[Post-Merge]-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-[Post-Merge]-1, xxx".

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

## Summary by CodeRabbit

* **New Features**
  * Added support for returning top-k log probabilities per token in chat completions, enabling detailed token probability insights.
  * Enhanced API to optionally include top log probabilities in chat completion responses.
  * Extended request parameters to better control logprobs gathering and backend-specific behavior.

* **Bug Fixes**
  * Improved validation and handling of logprobs and top_logprobs parameters in chat completion requests.
  * Refined internal logprob state management for more accurate probability tracking.

* **Tests**
  * Added new integration and unit tests validating top-k logprobs functionality on PyTorch and TensorRT backends.
  * Consolidated and updated existing tests for logprobs and chat completions, removing redundant streaming tests.

* **Documentation**
  * Updated parameter validation logic and API behavior for logprobs and top_logprobs fields in chat completion requests.

### CODE REVIEW

File: tests/unittest/llmapi/apps/openai_server.py [hunk 1]
```diff
@@ -28,7 +28,8 @@ def __init__(self,
         args = ["--host", f"{self.host}", "--port", f"{self.port}"]
         if cli_args:
             args += cli_args
-        launch_cmd = ["trtllm-serve"] + [model] + args
+        launch_cmd = ["python3"] + ["-m"] + ["tensorrt_llm.commands.serve"
```
[the remark this answers is not in the document]
Comment (LinPoly) [reply tests/unittest/llmapi/apps/openai_server.py:31]: Thx, this is not intended, will revert this line, `trtllm-serve` is just fine

File: tensorrt_llm/executor/result.py [hunk 2]
```diff
@@ -78,7 +78,7 @@ class CompletionOutput:
         text (str): The generated output text. Defaults to "".
         token_ids (List[int], optional): The token ids of the generated output text. Defaults to [].
         cumulative_logprob (float, optional): The cumulative log probability of the generated output text. Defaults to None.
-        logprobs (TokenLogprobs, optional): The log probabilities of the top probability words at each position if the logprobs are requested. Defaults to None.
+        logprobs (TokenLogprobs | List[float], optional): The log probabilities of the top probability words at each position if the logprobs are requested. Defaults to None.
```
[the remark this answers is not in the document]
Comment (LinPoly) [reply tensorrt_llm/executor/result.py:81]: `List[float]` is for a kind of special case in TRT backend, `gather_generation_logits` can be false and TRT backend can still return logprob of the current token (_NOT top logprobs_), whose format is float. Theoretically we can wrap this into `TokenLogprobs`, but I am not sure how much effort it would take, some C++ modification maybe needed. And I think we should allow user to do so, `gather_generation_logits` has its cost and it is reasonable to just want to get the logprob.

File: tensorrt_llm/serve/openai_protocol.py [hunk 3]
```diff
@@ -558,10 +559,16 @@ def to_sampling_params(self) -> SamplingParams:
...
         )
+        if self.logprobs:
+            logprobs = 1 if not self.top_logprobs else self.top_logprobs
+            if backend == "pytorch":
+                sampling_params.logprobs = logprobs
+            else:
+                if gather_generation_logits:
+                    sampling_params.logprobs = logprobs
+                else:
+                    sampling_params._return_log_probs = True
```
[the remark this answers is not in the document]
Comment (LinPoly) [reply tensorrt_llm/serve/openai_protocol.py:571]: Thx, this should be fixed

[the remark this answers is not in the document]
Comment (LinPoly) [reply tensorrt_llm/serve/openai_protocol.py:575]: Checking added

File: tensorrt_llm/executor/result.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (syuoni) [reply tensorrt_llm/executor/result.py:81]: I see. Thanks for explanation.

**Review by syuoni — APPROVED** [review]

**Review by QiJune — APPROVED** [review]
